File: ui/android-app/app/src/main/java/org/aerobag/app/RouteLabelLayout.kt
// SPDX-FileCopyrightText: 2026 Aerobag contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package org.aerobag.app

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import kotlin.math.*

// Pointer-rate mirror of app_core::ui_geometry::route_labels; shared conformance
// vectors fence placement, clipping and the protected-label collision fallback.
internal data class RouteLabelLayout(val anchor: Offset, val baseline: Offset,
    val bounds: Rect, val leader: Offset?)
internal data class RouteLabelCandidate(val from: Offset, val to: Offset, val label: String, val important: Boolean)

internal fun routeLabelIndices(legs: List<RouteLabelCandidate>, width: Float, height: Float): List<Int> {
    val result = mutableListOf<Int>()
    var index = 0
    while (index < legs.size) {
        val first = index++
        if (!legs[first].important) { result.add(first); continue }
        while (index < legs.size && legs[index].important && legs[index].label == legs[first].label &&
            legs[index-1].to == legs[index].from) index++
        var best: Int? = null; var length = -1f
        for (member in first until index) {
            val leg = legs[member]
            val interval = visibleInterval(leg.from,leg.to,width,height) ?: continue
            val visibleLength = (leg.to-leg.from).getDistance()*(interval.second-interval.first)
            if (visibleLength > length+1e-4f) { length = visibleLength; best = member }
        }
        best?.let { result.add(it) }
    }
    return result.sortedByDescending { legs[it].important }
}

internal fun routeLabelBounds(baseline: Offset, textWidth: Float) = Rect(
    baseline.x-textWidth/2f-4f, baseline.y-17f, baseline.x+textWidth/2f+4f, baseline.y+7f)

private fun visibleInterval(from: Offset, to: Offset, width: Float, height: Float): Pair<Float,Float>? {
    var start = 0f; var end = 1f
    for ((origin, delta, limit) in listOf(Triple(from.x,to.x-from.x,width), Triple(from.y,to.y-from.y,height))) {
        if (abs(delta) < 1e-9f) {
            if (origin < 0f || origin > limit) return null
        } else {
            val a = -origin/delta; val b = (limit-origin)/delta
            start = max(start,min(a,b)); end = min(end,max(a,b))
        }
    }
    return if (start <= end) Pair(start,end) else null
}

/** All dimensions are in density-independent pixels in the displayed map frame. */
internal fun routeLabelLayout(from: Offset, to: Offset, width: Float, height: Float,
    textWidth: Float, occupied: List<Rect>, important: Boolean): RouteLabelLayout? {
    if (width < 32f || height < 40f) return null
    val (start,end) = visibleInterval(from,to,width,height) ?: return null
    val fraction = if (important) (start+end)/2f else 0.5f
    val anchor = from+(to-from)*fraction
    val preferred = anchor-Offset(0f,10f)
    fun overlap(bounds: Rect) = occupied.sumOf { other ->
        (max(0f,min(bounds.right,other.right)-max(bounds.left,other.left)) *
            max(0f,min(bounds.bottom,other.bottom)-max(bounds.top,other.top))).toDouble()
    }
    var baseline = preferred
    if (important) {
        val inset = min(textWidth/2f+12f,width/2f); val stepX = max(textWidth+12f,32f)
        val columns = ceil(width/stepX).toInt(); val rows = ceil(height/32f).toInt()
        var bestOverlap = Double.POSITIVE_INFINITY; var bestDistance = Float.POSITIVE_INFINITY
        for (row in -rows..rows) for (column in -columns..columns) {
            val candidate = Offset((preferred.x+column*stepX).coerceIn(inset,width-inset),
                (preferred.y+row*32f).coerceIn(25f,height-15f))
            val area = overlap(routeLabelBounds(candidate,textWidth))
            val distance = (candidate-preferred).getDistanceSquared()
            if (area < bestOverlap || (area == bestOverlap && distance < bestDistance)) {
                bestOverlap = area; bestDistance = distance; baseline = candidate
            }
        }
    } else {
        val bounds = routeLabelBounds(baseline,textWidth)
        if (bounds.left < 8f || bounds.right > width-8f || bounds.top < 8f || bounds.bottom > height-8f || overlap(bounds)>0) return null
    }
    val bounds = routeLabelBounds(baseline,textWidth)
    val leader = if (abs(baseline.x-preferred.x)>1f || abs(baseline.y-preferred.y)>1f)
        Offset(anchor.x.coerceIn(bounds.left,bounds.right),anchor.y.coerceIn(bounds.top,bounds.bottom)) else null
    return RouteLabelLayout(anchor,baseline,bounds,leader)
}

File: ui/web-app/src/domain/routeLabelLayout.ts
// SPDX-FileCopyrightText: 2026 Aerobag contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

// Pointer-rate mirror of app_core::ui_geometry::route_labels; shared conformance
// vectors fence placement, clipping and the protected-label collision fallback.
type Point = {x: number; y: number};
export type LabelRect = {left: number; top: number; right: number; bottom: number};
export type RouteLabelLayout = {anchor: Point; baseline: Point; bounds: LabelRect; leader: Point | null};
type RouteLabelCandidate = {from: Point; to: Point; label: string; important: boolean};

export function routeLabelIndices(legs: readonly RouteLabelCandidate[], width: number, height: number): number[] {
  const result: number[] = [];
  let index = 0;
  while (index < legs.length) {
    const first = index++;
    if (!legs[first].important) { result.push(first); continue; }
    while (index < legs.length && legs[index].important && legs[index].label === legs[first].label &&
      legs[index-1].to.x === legs[index].from.x && legs[index-1].to.y === legs[index].from.y) index++;
    let best: number | null = null, length = -1;
    for (let member=first; member<index; member++) {
      const leg = legs[member], interval = visibleInterval(leg.from,leg.to,width,height);
      if (interval) {
        const visibleLength = Math.hypot(leg.to.x-leg.from.x,leg.to.y-leg.from.y)*(interval[1]-interval[0]);
        if (visibleLength > length+1e-4) { length = visibleLength; best = member; }
      }
    }
    if (best != null) result.push(best);
  }
  return result.sort((a,b) => Number(legs[b].important)-Number(legs[a].important));
}

export function routeLabelBounds(baseline: Point, textWidth: number): LabelRect {
  return {left: baseline.x-textWidth/2-4, top: baseline.y-17,
    right: baseline.x+textWidth/2+4, bottom: baseline.y+7};
}

function visibleInterval(from: Point, to: Point, width: number, height: number): [number,number] | null {
  let start = 0, end = 1;
  for (const [origin, delta, limit] of [[from.x, to.x-from.x, width], [from.y, to.y-from.y, height]]) {
    if (Math.abs(delta) < 1e-9) {
      if (origin < 0 || origin > limit) return null;
    } else {
      const a = -origin/delta, b = (limit-origin)/delta;
      start = Math.max(start, Math.min(a,b)); end = Math.min(end, Math.max(a,b));
    }
  }
  return start <= end ? [start,end] : null;
}

export function routeLabelLayout(from: Point, to: Point, width: number, height: number,
  textWidth: number, occupied: readonly LabelRect[], important: boolean): RouteLabelLayout | null {
  if (width < 32 || height < 40) return null;
  const interval = visibleInterval(from,to,width,height);
  if (!interval) return null;
  const [start,end] = interval;
  const fraction = important ? (start+end)/2 : 0.5;
  const anchor = {x:from.x+(to.x-from.x)*fraction, y:from.y+(to.y-from.y)*fraction};
  const preferred = {x:anchor.x, y:anchor.y-10};
  const overlap = (bounds: LabelRect) => occupied.reduce((sum, other) => sum +
    Math.max(0, Math.min(bounds.right,other.right)-Math.max(bounds.left,other.left)) *
    Math.max(0, Math.min(bounds.bottom,other.bottom)-Math.max(bounds.top,other.top)), 0);
  let baseline = preferred;
  if (important) {
    const inset = Math.min(textWidth/2+12, width/2), stepX = Math.max(textWidth+12, 32);
    const columns = Math.ceil(width/stepX), rows = Math.ceil(height/32);
    let bestOverlap = Infinity, bestDistance = Infinity;
    for (let row=-rows; row<=rows; row++) for (let column=-columns; column<=columns; column++) {
      const candidate = {x:Math.max(inset, Math.min(width-inset, preferred.x+column*stepX)),
        y:Math.max(25, Math.min(height-15, preferred.y+row*32))};
      const area = overlap(routeLabelBounds(candidate, textWidth));
      const distance = (candidate.x-preferred.x)**2+(candidate.y-preferred.y)**2;
      if (area < bestOverlap || (area === bestOverlap && distance < bestDistance)) {
        bestOverlap = area; bestDistance = distance; baseline = candidate;
      }
    }
  } else {
    const bounds = routeLabelBounds(baseline, textWidth);
    if (bounds.left < 8 || bounds.right > width-8 || bounds.top < 8 || bounds.bottom > height-8 || overlap(bounds)>0) return null;
  }
  const bounds = routeLabelBounds(baseline, textWidth);
  const leader = Math.abs(baseline.x-preferred.x)>1 || Math.abs(baseline.y-preferred.y)>1
    ? {x:Math.max(bounds.left, Math.min(bounds.right, anchor.x)), y:Math.max(bounds.top, Math.min(bounds.bottom, anchor.y))} : null;
  return {anchor, baseline, bounds, leader};
}

let measurement: CanvasRenderingContext2D | null = null;
export function measureRouteLabel(text: string, important = false): number {
  measurement ??= document.createElement("canvas").getContext("2d");
  if (!measurement) throw new Error("2D canvas context is required to measure route labels");
  measurement.font = `${important ? 800 : 600} 13px system-ui`;
  return measurement.measureText(text).width;
}
